src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/ProjectExtensions.cs

```diff
@@ -58,7 +59,110 @@ public static async ValueTask<bool> AppliesToProjectAsync(this Type type, IProje
                 }
             }
 
+            // Check whether the type has an [DiagnosticAnalyzer] attribute
+            // If one exists, the type only applies to the project if the language matches
+            if (!AppliesToLanguage(type, project))
+            {
+                return false;
+            }
+
             return true;
         }
+
+        /// <summary>
+        /// Checks to see if the type should be filtered out of the user's view.
+        /// </summary>
+        /// <param name="type">a type that represents an upgrade-assistant feature.</param>
+        /// <param name="project">the project currently being upgraded.</param>
+        /// <returns>true if the object is described as matching the project's language. Objects not defning the language they support will be true as default.</returns>
+        private static bool AppliesToLanguage(Type type, IProject project)
+        {
+            // check to see if this is a code fixer
+            if (DoesCodeFixerFilterThisFeature(type, project))
+            {
+                return false;
+            }
+
+            // check to see if this is a configuration step or other feature using the ApplicableLanguageAttribute
+            if (DoesApplicableLanguageFilterThisFeature(type, project))
+            {
+                return false;
+            }
+
+            // if the attribute is not applied then this type defaults to opt into language support
+            return true;
+        }
+
+        /// <summary>
+        /// Checks to see if the type is a codefixer that should be filtered out of the user's view.
+        /// </summary>
+        /// <param name="type">a type that represents an upgrade-assistant feature decorated with <see cref="ExportCodeFixProviderAttribute"/>.</param>
+        /// <param name="project">the project currently being upgraded.</param>
+        /// <returns>true if the codefixer should be hidden from the user. Will return false for non-codefixer objects.</returns>
+        private static bool DoesCodeFixerFilterThisFeature(Type type, IProject project)
+        {
+            var analyzerAttr = type.CustomAttributes.FirstOrDefault(a => a.AttributeType.FullName.Equals(typeof(ExportCodeFixProviderAttribute).FullName, StringComparison.Ordinal));
+            if (analyzerAttr is not null)
+            {
+                var applicableLanguage = analyzerAttr.ConstructorArguments.First().Value as string;
```

Comment:
This only checks the attribute's first argument, but the attribute also takes a `params string[]` of "additional languages". We should check whether *any* of the languages match the project's language.

Reply:
Once we added support for multiple languages to codefixers it made sense to just do the appicableLanguageAttr too

Reply:
This one feels different to me than ApplicableLanguageAttr. The ability to specify multiple languages on a code fix provider is already present, so I think we should respect that additional languages that are specified. In fact, I'm not sure if it's possible to create a code fix provider *without* specifying language, so a multi-language code fix would have to include a list of languages (unlike other components that could just leave `[ApplicableLanguage]` off in those cases.

Reply:
agreed